I believe the root cause is that the `integer_atomics` feature must be enabled to use AtomicU/I128. AtomicU/I128 isn't present on some platforms, so I am guessing that is why no one has run into this yet. This PR shouldn't limit functionality on any platforms as the `integer_atomics` feature isn't enabled in any scenarios on `main`, this PR just enables it if the platform has 128b atomics. A simpler change might be to just always enable the `integer_atomics` feature, even on platforms where it isn't needed.


```
error[E0658]: use of unstable library feature 'integer_atomics'
  --> C:\Users\kevin\.cargo\registry\src\index.crates.io-6f17d22bba15001f\const-type-layout-0.3.2\src\impls\core\sync\atomic.rs:9:36
   |
9  |           unsafe impl TypeLayout for core::sync::atomic::$at {
   |                                      ^
...
39 | / impl_atomic_int_layout! {
40 | |     AtomicBool (1:"8") => u8 => false,
41 | |     AtomicI8 (1:"8") => i8 => 0, AtomicI16 (2:"16") => i16 => 0,
42 | |     AtomicI32 (4:"32") => i32 => 0, AtomicI64 (8:"64") => i64 => 0,
...  |
45 | |     AtomicI128 (16:"128") => i128 => 0, AtomicU128 (16:"128") => u128 => 0
46 | | }
   | |_- in this macro invocation
   |
   = note: see issue #99069 <https://github.com/rust-lang/rust/issues/99069> for more information
   = help: add `#![feature(integer_atomics)]` to the crate attributes to enable
   = note: this compiler was built on 2024-07-20; consider upgrading it if it is out of date
   = note: this error originates in the macro `impl_atomic_int_layout` (in Nightly builds, run with -Z macro-backtrace for more info)

```

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -48,6 +48,7 @@ derive = ["dep:const-type-layout-derive"]
 serde = ["dep:serde"]
 
 impl-atomics = []
+impl-atomics-128 = ["impl-atomics"]
 impl-never = []
 impl-sync-exclusive = []
 impl-sync-unsafe-cell = []

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -136,10 +136,7 @@ r#"TypeLayoutInfo {
 // optional feature-gated features
 // https://github.com/rust-lang/rust/issues/94039
 #![cfg_attr(feature = "impl-atomics", feature(cfg_target_has_atomic))]
-#![cfg_attr(
-    all(feature = "impl-atomics", target_has_atomic_load_store = "128"),
-    feature(integer_atomics)
-)]
+#![cfg_attr(feature = "impl-atomics-128", feature(integer_atomics))]
 #![cfg_attr(feature = "impl-never", feature(never_type))]
 #![cfg_attr(feature = "impl-sync-exclusive", feature(exclusive_wrapper))]
 #![cfg_attr(feature = "impl-sync-unsafe-cell", feature(sync_unsafe_cell))]

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -135,7 +135,7 @@ r#"TypeLayoutInfo {
 #![cfg_attr(doc, feature(doc_auto_cfg))]
 // optional feature-gated features
 // https://github.com/rust-lang/rust/issues/94039
-#![feature(cfg_target_has_atomic)]
+#![cfg_attr(feature = "impl-atomics", feature(cfg_target_has_atomic))]
 #![cfg_attr(
     all(feature = "impl-atomics", target_has_atomic_load_store = "128"),
     feature(integer_atomics)

